src/v/config/constraint_method_generators.cc

```diff
@@ -0,0 +1,561 @@
+#include "config/configuration.h"
+#include "model/fundamental.h"
+#include "vassert.h"
+
+#include <seastar/util/log.hh>
+
+#include <algorithm>
+
+namespace config {
+inline ss::logger logger{"constraints"};
+
+namespace {
+
+template<typename RangeMinT, typename T>
+bool less_than_min(const std::optional<RangeMinT> min, const T& topic_val) {
+    return min && topic_val < static_cast<T>(*min);
+}
+
+template<typename RangeMaxT, typename T>
+bool greater_than_max(const std::optional<RangeMaxT> max, const T& topic_val) {
+    return max && topic_val > static_cast<T>(*max);
+}
+
+template<typename RangeT, typename T>
+bool is_invalid_range(const range_values<RangeT>& range, const T& topic_val) {
+    return less_than_min<RangeT, T>(range.min, topic_val)
+           || greater_than_max<RangeT, T>(range.max, topic_val);
+}
+
+template<typename RangeT>
+range_values<RangeT>
+get_range(const constraint_args& args, const model::topic& topic) {
+    // NOTE to reviewers: Not using visit here or else need to check for
+    // signed/unsigned int comparison which is avoidable.
+    try {
+        return std::get<range_values<RangeT>>(args.flags);
+    } catch (const std::bad_variant_access& ex) {
+        // This should not happen since the variant is hardcoded in
+        // constraint_types.cc, but just-in-case, let's catch and log.
+        logger.error(
+          "Constraints failure[min/max not found]: topic property {}.{}",
+          topic(),
+          args.name);
+        return range_values<RangeT>{};
+    }
+}
+
+template<typename RangeT, typename T>
+bool validate_range(
+  const T& topic_val, const model::topic& topic, const constraint_args& args) {
+    range_values<RangeT> range = get_range<RangeT>(args, topic);
+    if (is_invalid_range<RangeT, T>(range, topic_val)) {
+        logger.error(
+          "Constraints failure[value out-of-range]: topic property "
+          "{}.{}, value {}",
+          topic(),
+          args.name,
+          topic_val);
+        return false;
+    }
+
+    return true;
+}
+
+template<typename RangeT, typename T>
+std::optional<T> get_min_max(const range_values<RangeT>& range, T& topic_val) {
+    if (less_than_min<RangeT, T>(range.min, topic_val)) {
+        return std::make_optional(static_cast<T>(*range.min));
+    }
+
+    if (greater_than_max<RangeT, T>(range.max, topic_val)) {
+        return std::make_optional(static_cast<T>(*range.max));
+    }
+
+    return std::nullopt;
+}
+
+template<typename RangeT, typename T>
+void maybe_clamp_range(
+  T& topic_val, const model::topic& topic, const constraint_args& args) {
+    range_values<RangeT> range = get_range<RangeT>(args, topic);
+    auto min_max = get_min_max<RangeT, T>(range, topic_val);
+    if (min_max) {
+        logger.warn(
+          "Overwriting topic property {}.{} to the constraint min/max",
+          topic(),
+          args.name);
+        topic_val = *min_max;
+    }
+}
+
+template<typename T>
+bool match_cluster_property(
```

Comment:
Ok I think I see what's going on with this function.  This validates if the value of the topic property matches the value of the cluster property.  nit: Maybe change name of function to like `match_cluster_property_value`?